td-shim-tools/src/bin/td-payload-reference-calculator/main.rs

```diff
@@ -31,6 +43,88 @@ fn param(param: &str, size: &str) -> Result<String> {
     padding_digest(param, siz)
 }
 
+fn qemu(path: &str, size: &str) -> Result<String> {
+    let path = Path::new(path).to_path_buf();
+    let siz = parse::<u64>(size)?;
+    let file_size = std::fs::metadata(&path)?.len();
+    if file_size > siz {
+        bail!("File size should be less than `kernel-size`");
+    }
+    let buf = std::fs::read(path)?;
+    qemu_patch(buf)
+}
+
+fn qemu_patch(mut buf: Vec<u8>) -> Result<String> {
+    // patching type_of_loader @0x210
+    buf[0x210] = 0xb0;
+
+    // patching loadflags @0x211
+    buf[0x211] = 0x81;
+
+    //patching heap_end_ptr @0x224 cmdline_addr - real_addr - 0x200 = 0xfe00
+    buf[0x224] = 0x00;
+    buf[0x225] = 0xfe;
+
+    //patching cmd_line_ptr @0x228 cmdline_addr = 0x20000
+    buf[0x228] = 0x00;
+    buf[0x229] = 0x00;
+    buf[0x22A] = 0x02;
+
+    let mut hasher = sha2::Sha384::new();
+    let image = buf.clone();
+    let (number_of_region_entry, regions_base, regions_size) = get_image_regions(image);
+
+    for index in 0..number_of_region_entry {
+        hasher.update(&buf[regions_base[index]..regions_base[index] + regions_size[index]]);
+    }
+
+    let res = hasher.finalize();
+    Ok(hex::encode(res))
+}
+
+fn get_image_regions(buf: Vec<u8>) -> (usize, Vec<usize>, Vec<usize>) {
+    // These 3 regions are known.
+    let mut number_of_region_entry = 3;
+    let mut regions_base = vec![
+        IMGAE_BEGIN_CHECKSUM_ADDR,
+        IMGAE_CERT_TABLE_ADDR,
+        IMGAE_HEADERS_ADDR,
+    ];
+    let mut regions_size = vec![
+        IMGAE_BEGIN_CHECKSUM_SIZE,
+        IMGAE_CERT_TABLE_SIZE,
+        IMGAE_HEADERS_SIZE,
+    ];
+
+    // Refer to https://learn.microsoft.com/en-us/windows/win32/debug/pe-format, the NumberOfSections' offset is 2
+    let pe_addr =
+        ((buf[IMAGE_PE_ADDR + 1] as u16) << 8) | (buf[IMAGE_PE_ADDR] as u16 + PE_SIGNATURE_SIZE);
```

Comment:
This calculation confused me. This field you're grabbing seems to be `e_lfanew`. I'm having trouble finding its size. Do you have a link? It may actually be 32 bits (e.g. [here](https://0xrick.github.io/win-internals/pe3/) or [here](https://wiki.osdev.org/MZ)), but you're taking 16 bits. (Maybe in practice it's only ever 16 bits?) Also, I wasn't sure why you added `PE_SIGNATURE_SIZE` to this. Don't you just want the value at 0x3c?

Reply:
Thank you for your correction!
1. It should be 32 bits, I've modified it.
2.  Here the link for coff-file-header: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#signature-image-only
I want to get the address of coff-file-header and it starts at `pe_addr` + `PE_SIGNATURE_SIZE`. The variables should be renamed, and I've modified it.